Kentor.AuthServices/Internal/PendingAuthnRequests.cs

```diff
@@ -1,40 +1,94 @@
-﻿using System;
-using System.Collections.Generic;
+﻿using System;
+using System.IdentityModel.Metadata;
+using System.IdentityModel.Services;
 using System.IdentityModel.Tokens;
-using System.Linq;
-using System.Text;
-using System.Threading.Tasks;
-
+using System.Text;
+using System.Web;
+using System.Web.Security;
+using Kentor.AuthServices.Configuration;
+
 namespace Kentor.AuthServices.Internal
 {
     static class PendingAuthnRequests
-    {
-        private static readonly Dictionary<Saml2Id, StoredRequestState> pendingAuthnRequest = new Dictionary<Saml2Id, StoredRequestState>();
-
-        internal static void Add(Saml2Id id, StoredRequestState idp)
-        {
-            lock (pendingAuthnRequest)
-            {
-                if (pendingAuthnRequest.ContainsKey(id))
-                {
-                    throw new InvalidOperationException("AuthnRequest id can't be reused.");
-                }
-                pendingAuthnRequest.Add(id, idp);
-            }
-        }
-
-        internal static bool TryRemove(Saml2Id id, out StoredRequestState idp)
-        {
-            lock (pendingAuthnRequest)
-            {
-                if (id != null && pendingAuthnRequest.ContainsKey(id))
-                {
-                    idp = pendingAuthnRequest[id];
-                    return pendingAuthnRequest.Remove(id);
-                }
-                idp = null;
-                return false;
-            }
-        }
+    {
+      private const string CookieName = "AuthNRequest";
+      private const string CookieIdAttribute = "id";
+      private const string CookieIdpIdAttribute = "idpId";
+      private const string Purpose = "Initiator Verification";
+
+      internal static void Add(Saml2Id id, StoredRequestState idp)
+      {
+        var cookie = GetNewCookie();
+        cookie.Values[CookieIdAttribute] = Encrypt(id.Value);
+        cookie.Values[CookieIdpIdAttribute] = Encrypt(idp.Idp.Id);

```

Comment:
Here ReturnUrl seems to be lost?
RelayData too.
Serialization of the whole StoredRequestState object might be a better solution?